x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx

```diff
@@ -0,0 +1,219 @@
+  refreshPaused: boolean;
+  refreshInterval: number;
+}
+
+// TODO this can be removed when EUI PR lands: https://github.com/elastic/eui/pull/1574
+declare module '@elastic/eui' {
+  interface OnTimeChangeProps {
+    start: string;
+    end: string;
+    isInvalid: boolean;
+    isQuickSelection: boolean;
+  }
+
+  interface OnRefreshChangeProps {
+    isPaused: boolean;
+    refreshInterval: number;
+  }
+
+  export type EuiSuperDatePickerProps = CommonProps & {
+    start?: string;
+    end?: string;
+    isPaused?: boolean;
+    refreshInterval?: number;
+    onTimeChange: (props: OnTimeChangeProps) => void;
+    onRefreshChange?: (props: OnRefreshChangeProps) => void;
+    showUpdateButton?: boolean;
+  };
+  export const EuiSuperDatePicker: React.SFC<EuiSuperDatePickerProps>;
+}
+
+const APM_DEFAULT_TIME_OPTIONS: DatePickerParams = {
+  rangeFrom: 'now-24h',
+  rangeTo: 'now',
+  refreshPaused: true,
+  refreshInterval: 0
+};
+
+class DatePickerComponent extends React.Component<Props> {
+  public state = {
+    from: '',
+    to: '',
+    isPaused: true,
+    refreshInterval: 0
+  };
+  public refreshTimeoutId = 0;
+
+  public getParamsFromSearch = memoize((search: string) => {
+    const query = toQuery(search);
+    const converted = {
+      ...query,
+      refreshPaused:
+        query.refreshPaused === 'true' || query.refreshPaused === true,
+      refreshInterval:
+        typeof query.refreshInterval === 'undefined'
+          ? undefined
+          : Number(query.refreshInterval)
+    };
+    return {
+      ...APM_DEFAULT_TIME_OPTIONS,
+      ...converted
+    } as DatePickerParams;
+  });
+
+  public componentDidMount() {
+    this.dispatchTimeRangeUpdate();
+    this.updateRefresh();
+  }
+
+  public componentWillUnmount() {
+    this.clearRefreshTimeout();
+  }
+
+  public componentDidUpdate(prevProps: Props) {
+    const currentParams = this.getParamsFromSearch(this.props.location.search);
+    const previousParams = this.getParamsFromSearch(prevProps.location.search);
+    if (
+      currentParams.rangeFrom !== previousParams.rangeFrom ||
+      currentParams.rangeTo !== previousParams.rangeTo
+    ) {
+      this.dispatchTimeRangeUpdate();
+    }
+
+    if (
+      currentParams.refreshPaused !== previousParams.refreshPaused ||
+      currentParams.refreshInterval !== previousParams.refreshInterval
+    ) {
+      this.updateRefresh();
+    }
+  }
+
+  public dispatchTimeRangeUpdate() {
+    const { rangeFrom, rangeTo } = this.getParamsFromSearch(
+      this.props.location.search
+    );
+    const parsed = {
+      from: datemath.parse(rangeFrom),
+      to: datemath.parse(rangeTo)
+    };
+    if (!parsed.from || !parsed.to) {
+      return;
+    }
+    const min = parsed.from.toISOString();
+    const max = parsed.to.toISOString();
+    this.props.dispatchUpdateTimePicker({ min, max });
+  }
+
+  public clearRefreshTimeout() {
+    if (this.refreshTimeoutId) {
+      window.clearTimeout(this.refreshTimeoutId);
```

Comment:
Explicitly using `window` here and for `window.setTimeout` below because Typescript otherwise assumes you are using the NodeJS implementation which returns a special "NodeJS.Timeout" typed value rather than just a number. 

We could alternatively use "as number" in a few places to tell TS what we want, but this seemed ok since Jest runs tests with JSDom.

Reply:
👍